src/System.Private.ServiceModel/src/System/ServiceModel/Channels/Connection.cs

```diff
@@ -31,6 +31,62 @@ public interface IConnection
         int EndRead();
     }
 
+    public static class IConnectionExtensions
+    {
+        public static async Task WriteAsync(this IConnection connection, byte[] buffer, int offset, int size, bool immediate, TimeSpan timeout)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+            var result = connection.BeginWrite(buffer, offset, size, immediate, timeout, OnIoComplete, tcs);
+            if (result == AsyncCompletionResult.Completed)
+            {
+                tcs.SetResult(true);
+            }
+
+            await tcs.Task;
+            connection.EndWrite();
+        }
+
+        public static async Task<int> ReadAsync(this IConnection connection, int offset, int size, TimeSpan timeout)
```

Comment:
This overload is not only used by the other overload only. It's used by ConnectionUpgradeHelper.DecodeFramingFaultAsync as well.